### PR TITLE
Matchmaking queue ban.

### PR DESCRIPTION
**Pull request by y4ure**

### CODE REVIEW

File: driftbase/api/matches.py [hunk 1]
```diff
@@ -886,6 +886,13 @@ def patch(self, args, match_id, player_id):
             log.info(f"player {player_id} not found in match {match_id}. Aborting.")
             abort(http_client.NOT_FOUND)
 
+        if args["status"] == "banned" and match_player.status != "banned":
+            log.info("Player %s is banned from battle %s", player_id, match_id)
+            log_match_event(match_id, player_id,"gameserver.match.player_banned")
+            current_app.extensions["messagebus"].publish_message("match", {
+                "event": "match_player_banned", "match_id": match_id, "player_id": player_id
+            })
```
Comment (y4ure) [inline driftbase/api/matches.py:894]: BattleServer will be updating the banned status of the player if
- In a ranked match
- Left the match and doesn't rejoin within rejoin time.
- Left the match and doesn't rejoin before the match ends.

File: driftbase/api/matches.py [hunk 2]
```diff
@@ -586,7 +586,7 @@ def put(self, args, match_id):
 
                 if new_status == "started":
                     match.start_date = utcnow()
-                elif new_status == "completed":
+                elif new_status in ("completed", "ended"):
```
Comment (y4ure) [inline driftbase/api/matches.py:589]: I see "end" in match active checks but is missing here.

Fixing battle server so aborted servers will have match_status set to "ended".

File: driftbase/resources/flexmatch.py [hunk 3]
```diff
@@ -9,8 +9,12 @@
...
     "aws_gamelift_role": "",
     "valid_regions": ["eu-west-1"],
-    "max_rejoin_time_seconds": 2 * 60,
-    "backfill_ticket_pattern": "^BackFill--.*"  # This is highly tenant specific; Perseus/TMA server issues backfill tickets with this prefix, but there's no rule here"""
+    "max_rejoin_time_seconds": 90,
+    "backfill_ticket_pattern": "^BackFill--.*",  # This is highly tenant specific; Perseus/TMA server issues backfill tickets with this prefix, but there's no rule here"""
+    "ban_times": {
+        "time_tiers_minutes": [6, 15, 30, 60, 12 * 60],
+        "expire_time_minutes": 24 * 60
+    }
```
Comment (y4ure) [inline driftbase/resources/flexmatch.py:17]: Escalating ban times, expires in an hour.

**Review by einarth — COMMENTED** [review]
Don't see anything wrong with the code, but I feel that there should be a test for the ban mechanic itself and its escalation.

**Comment by y4ure** [comment]
> Don't see anything wrong with the code, but I feel that there should be a test for the ban mechanic itself and its escalation.

Rewrote the logic to support separate banning for different matchmakers.
Updated unit test.

File: driftbase/flexmatch.py [hunk 4]
```diff
@@ -793,3 +807,69 @@ def __init__(self, user_message, debug_info):
         super().__init__(user_message, debug_info)
         self.msg = user_message
         self.debugs = debug_info
+
+class BanInfo(object):
+    test_store = None
```
[the remark this answers is not in the document]
Comment (y4ure) [reply driftbase/flexmatch.py]: Updated

**Pull request merged**